File: app/core/derived/node_view.py
"""NodeView derived-state appliers."""

from __future__ import annotations

import json
import sqlite3

from app.core.operation import Operation


def _json(value: object) -> str:
    return json.dumps(value)


def _maybe_json(value: object | None) -> str | None:
    if value is None:
        return None
    return json.dumps(value)


def apply_node_view_create(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply a ``nodeView.create`` operation."""
    payload = op.payload
    ts = op.envelope.timestamp.isoformat() if op.envelope.timestamp else None
    conn.execute(
        """
        INSERT OR REPLACE INTO node_view (
            id, workspace_id, node_id, name, view_type, order_index, is_default,
            active, shown_properties, group_by, view_mode, sort_entries,
            settings, query_ast, created_at, updated_at
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            payload["viewId"],
            op.envelope.workspace_id,
            payload["nodeId"],
            payload["name"],
            payload["viewType"],
            payload.get("orderIndex", 0),
            1 if payload.get("isDefault") else 0,
            1,
            _json(payload.get("shownProperties", [])),
            _maybe_json(payload.get("groupBy")),
            payload.get("viewMode"),
            _json(payload.get("sortEntries", [])),
            _json(payload.get("settings", {})),
            _maybe_json(payload.get("queryAst")),
            ts,
            ts,
        ),
    )


def apply_node_view_update(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply a ``nodeView.update`` operation.

    The payload stores the full updated record; mutable columns are overwritten
    so that ``null`` clears a field and absent fields do not change.
    """
    payload = op.payload
    view_id = payload["viewId"]
    ts = op.envelope.timestamp.isoformat() if op.envelope.timestamp else None

    columns: list[str] = []
    values: list[object] = []

    def add(column: str, value: object) -> None:
        columns.append(f"{column} = ?")
        values.append(value)

    if "name" in payload:
        add("name", payload["name"])
    if "orderIndex" in payload:
        add("order_index", payload["orderIndex"])
    if "isDefault" in payload:
        add("is_default", 1 if payload["isDefault"] else 0)
    if "shownProperties" in payload:
        add("shown_properties", _json(payload["shownProperties"]))
    if "groupBy" in payload:
        add("group_by", _maybe_json(payload["groupBy"]))
    if "viewMode" in payload:
        add("view_mode", payload["viewMode"])
    if "sortEntries" in payload:
        add("sort_entries", _json(payload["sortEntries"]))
    if "settings" in payload:
        add("settings", _json(payload["settings"]))
    if "queryAst" in payload:
        add("query_ast", _maybe_json(payload["queryAst"]))

    if not columns:
        return

    add("updated_at", ts)
    values.append(view_id)

    conn.execute(
        f"UPDATE node_view SET {', '.join(columns)} WHERE id = ?",
        values,
    )


def apply_node_view_delete(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply a ``nodeView.delete`` operation."""
    conn.execute(
        "DELETE FROM node_view WHERE id = ?",
        (op.payload["viewId"],),
    )


def apply_node_view_reorder(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply a ``nodeView.reorder`` operation."""
    payload = op.payload
    node_id = payload["nodeId"]
    view_type = payload["viewType"]
    ordered_ids = payload.get("orderedViewIds", [])
    for idx, view_id in enumerate(ordered_ids):
        conn.execute(
            """
            UPDATE node_view
            SET order_index = ?
            WHERE node_id = ? AND view_type = ? AND id = ?
            """,
            (idx, node_id, view_type, view_id),
        )


def delete_node_views_for_node(conn: sqlite3.Connection, node_id: str) -> None:
    """Remove all node_view rows belonging to ``node_id``."""
    conn.execute("DELETE FROM node_view WHERE node_id = ?", (node_id,))

File: frontend/src/core/derived/nodeView.ts
import { type Database } from 'sql.js';
import type { Operation } from '../types/operation';
import { queryOne } from '../db/sqlite';
import type { ChangeNotification } from './index';

function jsonOrNull(value: unknown): string | null {
  if (value === undefined || value === null) return null;
  return JSON.stringify(value);
}

function jsonOrDefault(value: unknown, defaultValue: unknown): string {
  if (value === undefined || value === null) return JSON.stringify(defaultValue);
  return JSON.stringify(value);
}

function applyNodeViewCreate(db: Database, op: Operation): ChangeNotification[] {
  const payload = op.payload as Record<string, unknown>;
  const now = new Date().toISOString();
  const nodeId = payload.nodeId as string;
  db.run(
    `INSERT OR REPLACE INTO node_view (
       id, workspace_id, node_id, name, view_type, order_index, is_default,
       active, shown_properties, group_by, view_mode, sort_entries,
       settings, query_ast, created_at, updated_at
     ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      payload.viewId as string,
      op.envelope.workspaceId,
      nodeId,
      payload.name as string,
      payload.viewType as string,
      (payload.orderIndex as number | undefined) ?? 0,
      payload.isDefault ? 1 : 0,
      1,
      jsonOrDefault(payload.shownProperties, []),
      jsonOrNull(payload.groupBy),
      (payload.viewMode as string | null | undefined) ?? null,
      jsonOrDefault(payload.sortEntries, []),
      jsonOrDefault(payload.settings, {}),
      jsonOrNull(payload.queryAst),
      now,
      now,
    ]
  );
  return [{ scope: 'node', nodeId }];
}

function applyNodeViewUpdate(db: Database, op: Operation): ChangeNotification[] {
  const payload = op.payload as Record<string, unknown>;
  const viewId = payload.viewId as string;
  const now = new Date().toISOString();

  const columns: string[] = [];
  const values: (string | number | null)[] = [];

  const addColumn = (name: string, value: string | number | null): void => {
    columns.push(`${name} = ?`);
    values.push(value);
  };

  if ('name' in payload) addColumn('name', payload.name as string);
  if ('orderIndex' in payload) addColumn('order_index', payload.orderIndex as number | null);
  if ('isDefault' in payload) addColumn('is_default', payload.isDefault ? 1 : 0);
  if ('shownProperties' in payload) addColumn('shown_properties', jsonOrDefault(payload.shownProperties, []));
  if ('groupBy' in payload) addColumn('group_by', jsonOrNull(payload.groupBy));
  if ('viewMode' in payload) addColumn('view_mode', (payload.viewMode as string | null | undefined) ?? null);
  if ('sortEntries' in payload) addColumn('sort_entries', jsonOrDefault(payload.sortEntries, []));
  if ('settings' in payload) addColumn('settings', jsonOrDefault(payload.settings, {}));
  if ('queryAst' in payload) addColumn('query_ast', jsonOrNull(payload.queryAst));

  if (columns.length === 0) return [];

  addColumn('updated_at', now);
  values.push(viewId);

  db.run(`UPDATE node_view SET ${columns.join(', ')} WHERE id = ?`, values);

  const row = queryOne<{ node_id: string }>(db, 'SELECT node_id FROM node_view WHERE id = ?', [viewId]);
  return row ? [{ scope: 'node', nodeId: row.node_id }] : [];
}

function applyNodeViewDelete(db: Database, op: Operation): ChangeNotification[] {
  const payload = op.payload as Record<string, unknown>;
  const viewId = payload.viewId as string;
  const row = queryOne<{ node_id: string }>(db, 'SELECT node_id FROM node_view WHERE id = ?', [viewId]);
  db.run('DELETE FROM node_view WHERE id = ?', [viewId]);
  return row ? [{ scope: 'node', nodeId: row.node_id }] : [];
}

function applyNodeViewReorder(db: Database, op: Operation): ChangeNotification[] {
  const payload = op.payload as Record<string, unknown>;
  const nodeId = payload.nodeId as string;
  const viewType = payload.viewType as string;
  const orderedIds = (payload.orderedViewIds as string[] | undefined) ?? [];
  for (let i = 0; i < orderedIds.length; i++) {
    db.run(
      'UPDATE node_view SET order_index = ? WHERE node_id = ? AND view_type = ? AND id = ?',
      [i, nodeId, viewType, orderedIds[i]]
    );
  }
  return [{ scope: 'node', nodeId }];
}

export function deleteNodeViewsForNode(db: Database, nodeId: string): void {
  db.run('DELETE FROM node_view WHERE node_id = ?', [nodeId]);
}

export function applyNodeViewOperation(db: Database, op: Operation): ChangeNotification[] {
  const { opType } = op.envelope;
  if (opType === 'nodeView.create') {
    return applyNodeViewCreate(db, op);
  }
  if (opType === 'nodeView.update') {
    return applyNodeViewUpdate(db, op);
  }
  if (opType === 'nodeView.delete') {
    return applyNodeViewDelete(db, op);
  }
  if (opType === 'nodeView.reorder') {
    return applyNodeViewReorder(db, op);
  }
  return [];
}
